### ISSUE TITLE
Would Zeek benefit from having customizable UIDs for both UID generation and output?

### ISSUE THREAD

**Issue by simonhf** [zeek/zeek#919]
Zeek uses 96bit [1] UIDs by default.

Which fits into 2 [2] 64bit unsigned integers.

The ascii UID is created by prefixing a 'C' or 'F' to 2 [3] base62 [4] strings.

The largest such UID is 18 bytes long, however, it's usual to see UIDs which are 1 or more bytes shorter because the length is not fixed or padded, e.g.:

```
Fo3qFkQKvrPkz0pF   <- example of 96bit UID as 16 bytes with F prefix
FnTiOes1yPs0S8y3h  <- example of 96bit UID as 17 bytes with F prefix
Fo22vP1oR9jwvOUnx2 <- example of 96bit UID as 18 bytes with F prefix
```

It would be nicer, more intention revealing, and/or more human friendly if the UIDs had a fixed length in the Zeek logs instead of variable length as at present? For example, padding the remaining digits with '0', e.g.?

```
Fo3qFkQKvrPkz0pF00 <- example of 96bit UID as 18 bytes with F prefix with '00' pad
FnTiOes1yPs0S8y3h0 <- example of 96bit UID as 18 bytes with F prefix with '0' pad
Fo22vP1oR9jwvOUnx2 <- example of 96bit UID as 18 bytes with F prefix with no pad
```

One possible way to implement this would be to modify UID::Set() [5] so that it calculates and caches the max bytes of `bits` passed to it. The calculation could be done by -- if no cached result exists -- get the ascii UID for the highest number of possible bits, e.g. if 96bits passed, then calculate the ascii UID for 2^96 -1 and then cache the resulting number of bytes which will be the longest UID string for 96bit UIDs. In this way an ascii UID result can be padded if it's less then the known calculated and cached maximum number of bytes.

[1] https://github.com/zeek/zeek/blob/master/scripts/base/init-bare.zeek#L5261
[2] https://github.com/zeek/zeek/blob/master/src/UID.h#L11
[3] https://github.com/zeek/zeek/blob/master/src/UID.cc#L39
[4] https://github.com/zeek/zeek/blob/master/src/util.cc#L684
[5] https://github.com/zeek/zeek/blob/master/src/UID.cc#L12

**Comment by 0xxon** [comment]
I already issued that opinion before - but just to have it here too - I think the current state that things are are ok.

The intention of the UID is to identify one connection (for a limited period of time). I don't see that being effected by adding more zeroes to the end.

Having variable lengths also comes with the small advantage that software so far does not depend on this being any specific length - which means that nothing should have problems when someone chooses a different length.

That being said - just my opinion.

**Comment by mavam** [comment]
There were some discussions in the past about having a customizable UID.

I don't have a good answer to the actual question, but when it comes to a Zeek-independent connection ID, then the [Community ID package](https://github.com/corelight/zeek-community-id) comes in handy. Suricata also computes an internal flow ID (albeit numeric) in addition to the Community ID. Depending on your use case, it may be enough, or even beneficial, to switch to that identifier.

**Comment by simonhf** [comment]
@mavam That would be a nice feature to be able to customize the UIDs. Not just to have control over the log formatting, but also regarding UID generation speed.

For example, in getting to know Zeek, I noticed that the file UID / FUID is kind of virtual in that it gets calculated potentially multiple times by MD5 hashing the same unique string every time it's needed, e.g. the unique string might be (note: "I recommend changing this." is the default salt appended to the unique string before hashing):

``
Analyzer::ANALYZER_HTTP1587145891.207408T11192.168.1.80:49365 > 207.152.124.122:80I recommend changing this.
``

If the file UID / FUID is needed a second time during Zeek processing then the unique string is recreated and re-hashed, meaning that the file UID / FUID is not stored anywhere at run-time AFAIK.

MD5 hashing is relatively old, and works byte by byte, and generates a 128bit hash of which 96bits are used by Zeek by default. If a Zeek user had the option to change to a different hash function then they could take advantage of newer algorithms for example which work on more than one byte at a time. I'm not advocating necessarily using xxHash but as an example, below, a 128bit xxHash appears to be about 30 times faster than the 128bit MD5 hash on my laptop:

```
$ gcc -O2 -o foo foo.c -lcrypto && ./foo
- hash64a=0x63ecd3a43772ad4 in 0.942264 seconds // <-- xxhash
- hash64b=0x182759938f63af00 in 27.857436 seconds // <-- md5
- md5 is 29.564362 times slower than xxhash
```

```
#include <stdio.h>
#include <locale.h>
#include <sys/mman.h>
#include <sys/time.h>
#include <openssl/evp.h>
#include <openssl/md5.h>

#include "xxHash/xxh3.h"

// install xxhash via: git clone https://github.com/Cyan4973/xxHash.git
// gcc -O2 -o foo foo.c -lcrypto && ./foo

double get_time_in_seconds(void) {
    struct timeval tv;

    gettimeofday(&tv, NULL);
    return (double)tv.tv_sec + 1.e-6 * (double)tv.tv_usec;
}

enum HashAlgorithm { Hash_MD5, Hash_SHA1, Hash_SHA224, Hash_SHA256, Hash_SHA384, Hash_SHA512 };

EVP_MD_CTX* hash_init(enum HashAlgorithm alg)
	{
	EVP_MD_CTX* c = EVP_MD_CTX_new();
	const EVP_MD* md;

	switch (alg)
		{
		case Hash_MD5:
#ifdef EVP_MD_CTX_FLAG_NON_FIPS_ALLOW
			/* Allow this to work even if FIPS disables it */
			EVP_MD_CTX_set_flags(c, EVP_MD_CTX_FLAG_NON_FIPS_ALLOW);
#endif
			md = EVP_md5();
			break;
		case Hash_SHA1:
			md = EVP_sha1();
			break;
		case Hash_SHA224:
			md = EVP_sha224();
			break;
		case Hash_SHA256:
			md = EVP_sha256();
			break;
		case Hash_SHA384:
			md = EVP_sha384();
			break;
		case Hash_SHA512:
			md = EVP_sha512();
			break;
		default:
			printf("Unknown hash algorithm passed to hash_init\n");
		}

	if ( ! EVP_DigestInit_ex(c, md, NULL) )
		printf("EVP_DigestInit failed\n");

	return c;
	}

void hash_update(EVP_MD_CTX* c, const void* data, unsigned long len)
	{
	if ( ! EVP_DigestUpdate(c, data, len) )
		printf("EVP_DigestUpdate failed\n");
	}

void hash_final(EVP_MD_CTX* c, u_char* md)
	{
	if ( ! EVP_DigestFinal(c, md, NULL) )
		printf("EVP_DigestFinal failed\n");

	EVP_MD_CTX_free(c);
	}

unsigned char* internal_md5(const unsigned char* data, unsigned long len, unsigned char* out)
	{
	static unsigned char static_out[MD5_DIGEST_LENGTH];

	if ( ! out )
		out = static_out; // use static array for return, see OpenSSL man page

	EVP_MD_CTX* c = hash_init(Hash_MD5);
	hash_update(c, data, len);
	hash_final(c, out);
	return out;
	}

void main(void) {
	char string[] = "Analyzer::ANALYZER_HTTP1587145891.207408T11192.168.1.80:49365 > 207.152.124.122:80I recommend changing this.";

	int hashes = 100000000;

	double t1a = get_time_in_seconds();
	uint64_t hash64a = 0;
	for (int i = 0; i < hashes; i ++) {
		XXH128_hash_t hash128 = XXH3_128bits_withSeed(&string[0], sizeof(string), i /* seed */);
		hash64a += hash128.low64;
	}
	double t2a = get_time_in_seconds();
	printf("- hash64a=0x%lx in %f seconds // <-- xxhash\n", hash64a, t2a - t1a);

	double t1b = get_time_in_seconds();
	uint64_t hash64b = 0;
	uint64_t hash[2];
	for (int i = 0; i < hashes; i ++) {
		internal_md5(&string[0], sizeof(string), (unsigned char *) &hash[0]);
		hash64b += hash[0];
	}
	double t2b = get_time_in_seconds();
	printf("- hash64b=0x%lx in %f seconds // <-- md5\n", hash64b, t2b - t1b);

	printf("- md5 is %f times slower than xxhash\n", (t2b - t1b) / (t2a - t1a));
}
```

**Comment by simonhf** [comment]
I had a quick look into the speed of the conversion of the 96bit UID to ascii.

Internally Zeek calls `uitoa_n()` twice to convert 2x 64bit ints to base62, and the resulting string gets written to a Zeek log file.

I tried one of the fastest avx / simd base64 implementations here [1], just for comparison. Obviously base64 'looks' horrible compared to base62...

I also tried conversion to hex instead of base62 because conversion to hex can be faster due to nibbles lining up better, etc. I copy and pasted the following code from here [2] to quickly test conversion to hex.

I hacked all the code into the Turbo Base64 example project. The results are as follows:

```
$ make && ./tb64app 
...
detected simd (id=60->'avx2')
- string_used_total=1600000000 in 0.912129 seconds <-- tb64enc()
- string_used_total=1700000000 in 3.673891 seconds <-- uitoa_n()
- val_total=1823103488 in 0.806929 seconds <-- nibble...()
- uitoa_n() is 4.027820 times slower than tb64enc()
- uitoa_n() is 4.552930 times slower than nibble...()
```

So on my laptop Turbo Base64 is about 4 times faster than `uitoa_n()` but it relies on special instructions which may not be available on the target CPU, and it produces 'horrible' looking base64 output instead of the base62 output Zeek currently uses... which would likely confuse some software which reads Zeek logs.

The fastest of the 3 methods tested is the conversion to hex which is 4.5 times faster than `uitoa_n()` but the 96bit hex output is 24 bytes instead of the base62 17 bytes. Seems like a potentially good deal, converting the 96bit UIDs to ascii 4.5 times faster, but outputting an extra 7 bytes...? What do you think?

So using something like xxhash instead of MD5 then the FUIDs could be created about 30 times faster. And converting the binary FUID to hex instead of base62 then the conversion could happen about 4.5 times faster.

It might also be possible to speed up the nibble to hex conversion further. Why? Because the functions from [2] work on 4 nibbles at a time with 32bit unsigned ints. In theory this could be converted to working on 8 nibbles at a time... The code is from 2015...

[1] https://github.com/powturbo/Turbo-Base64
[2] https://github.com/WojciechMula/toys/tree/master/conv_to_hex

**Comment by 0xxon** [comment]
Just a few thoughts:

the overhead of base62 is tiny in comparison to the other things we do. It probably is worth the cycles to save output.

At least UIDs use siphash, not MD5.

I am not quite sure if FUIDs use a separate codepath - but if they use normal hashing, they do actually use hmac_md5, not md5 (which has a different speed).

The place to look is the end of src/Hash.cc - we use siphash/hmac_md5 dependent on input size length. I am honestly not sure if there is a big reason to do this anymore - that is kind of historic. We could just switch everything to siphash.

xxhash is not a fair comparison - the hash we use should be cryptographically safe. @timwoj actually took a peek at this in the not to distant past - and we just exchanged siphash with a faster variant of it. Also - currently hashing does not seem to be a major performance choke point from what I remember.

**Comment by 0xxon** [comment]
As a small addition - I don't think the idea of being able to customize the internally used hash function is unreasonable.

**Comment by simonhf** [comment]
Interestingly, redis also uses siphash but a self-modified version [1] which it says is faster but "not believed to be as strong as the suggested 2-4 variant, but AFAIK there are not trivial attacks against this reduced-rounds version". The code comments  [1] say that the self-modified version of siphash "runs at the same speed as Murmurhash2, while the 2-4 variant slowed down Redis by a 4% figure more or less".

[1] https://github.com/antirez/redis/blob/unstable/src/siphash.c#L19

**Comment by 0xxon** [comment]
It turns out that there are indeed a few more places that use md5. I will take a stab at removing these - and also unifying the source a bit more to make swapping out of hash-functions easier.

**Comment by simonhf** [comment]
Yep, I noticed using cwrap [1] on Zeek that `internal_md5()` is called no less than 7,036 times even before `net_run()`.
[1] https://github.com/corelight/cwrap

**Comment by 0xxon** [comment]
I am going to close this. I think the speed issue has been resolved using #966, which basically removes any internal use of MD5 and HMAC-MD5, switching to highwayhash and siphash.

In the C++-source it is now also much easier to update the hash algorithms to something different.

Apart from that - I don't think that it is worth keeping the issue open anymore. I don't think there is clear consensus to change the output format for hashes. I agree that it would in principle be need to have an option to change the hash algorithms - but that seems rather non-urgent and probably won't have significant real-world speed impact. And - we don't need an open issue for this - if anyone wants to actively contribute this in the future we can just re-open, or go directly to a pull request.
